### PR TITLE
confdb: support list indexing paths in confdb-schema 

### PR DESCRIPTION
**Pull request by miguelpires**
8e8416432f03c663945cfc5cd3bd2f34486fea13 extends the dotted path validation to prevent the same placeholder name to be used more than once in a request
3703dc0e04bd93263866d8e843291e6239f571e8 supports paths with list indexes, both literal and placeholders (foo[1], foo[{n}])

### CODE REVIEW

File: confdb/confdb.go [hunk 1]
```diff
@@ -475,28 +523,73 @@ type validationOptions struct {
...
 
-	subkeys := strings.Split(path, ".")
+	subkeys, err := splitViewPath(path)
+	if err != nil {
+		return err
+	}
+
 	for _, subkey := range subkeys {
-		if subkey == "" {
+		if !validSubkey.MatchString(subkey) && !validIndexSubkey.MatchString(subkey) &&
```
Comment (andrewphelpsj) [inline confdb/confdb.go]: This condition is getting pretty big, maybe pull something out into a variable with a name?

File: confdb/confdb.go [hunk 2]
```diff
@@ -475,28 +523,73 @@ type validationOptions struct {
...
+		return nil, err
+	}
+
+	return subkeys, nil
 }
 
 // getPlaceholders returns the set of placeholders in the string or nil, if
 // there is none.
-func getPlaceholders(viewStr string) map[string]int {
+func getPlaceholders(path string, filter func(string) (string, bool)) (map[string]int, error) {
```
Comment (andrewphelpsj) [inline confdb/confdb.go]: Maybe mention here on this doc comment what filter is doing, filtering values and stripping our syntax from index and key placeholders.

File: confdb/confdb.go [hunk 3]
```diff
@@ -475,28 +523,73 @@ type validationOptions struct {
...
 			return errors.New("cannot have empty subkeys")
 		}
+		subkeys = append(subkeys, sb.String())
+		sb.Reset()
+		return nil
+	}
 
-		if !validSubkey.MatchString(subkey) && (!opts.allowPlaceholder || !validPlaceholder.MatchString(subkey)) {
-			return fmt.Errorf("invalid subkey %q", subkey)
+	for _, c := range path {
```
Comment (andrewphelpsj) [inline confdb/confdb.go:613]: Side note: we should consider looking into go's builtin fuzzing support to test some of this code for weird edge cases.

File: confdb/confdb.go [hunk 4]
```diff
@@ -1301,11 +1437,17 @@ func newViewRule(request, storage, accesstype string) (*viewRule, error) {
...
+		return nil, err
+	}
+
 	pathWriters := make([]storageWriter, 0, len(pathSubkeys))
 	for _, subkey := range pathSubkeys {
 		var patt storageWriter
-		if isPlaceholder(subkey) {
+		if isIndexPlaceholder(subkey) {
+			patt = indexPlaceholder(subkey[2 : len(subkey)-2])
+		} else if isPlaceholder(subkey) {
```
Comment (andrewphelpsj) [inline confdb/confdb.go]: Maybe rename to `isKeyPlaceholder`?

File: confdb/confdb.go [hunk 5]
```diff
@@ -1288,11 +1417,18 @@ func newViewRule(request, storage, accesstype string) (*viewRule, error) {
...
+		return nil, err
+	}
+
 	requestMatchers := make([]requestMatcher, 0, len(requestSubkeys))
 	for _, subkey := range requestSubkeys {
 		var patt requestMatcher
-		if isPlaceholder(subkey) {
+
+		if isIndexPlaceholder(subkey) {
+			patt = indexPlaceholder(subkey[2 : len(subkey)-2])
```
Comment (andrewphelpsj) [inline confdb/confdb.go]: We do this sub-slicing a lot, often right after checking the type of placeholder. Maybe worth making some sort of helper? Direct indexing always makes me nervous, so I like to keep it tightly coupled to the validation.

Comment (andrewphelpsj) [reply confdb/confdb.go]: I'm not quite sure how it'd look though.

File: confdb/confdb.go [hunk 6]
```diff
@@ -1408,18 +1554,28 @@ type placeholder string
...
+	matched.key[string(p)] = subkey
 	return true
 }
 
 // write writes the value from the placeholders map corresponding to this placeholder
 // key into the strings.Builder.
-func (p placeholder) write(sb *strings.Builder, placeholders map[string]string) error {
-	subkey, ok := placeholders[string(p)]
+func (p placeholder) write(sb *strings.Builder, matched *matchedPlaceholders, opts writeOpts) error {
+	subkey, ok := matched.key[string(p)]
```
Comment (andrewphelpsj) [inline confdb/confdb.go:1602]: matched.key is lazy-initialized above, is it possible to get here without initializing the map?

**Review by andrewphelpsj — COMMENTED** [review]
Thanks! A few small comments and questions.

File: confdb/schema_test.go [hunk 7]
```diff
@@ -2230,7 +2230,7 @@ func (*schemaSuite) TestSchemaAtNestedInArray(c *C) {
 	schema, err := confdb.ParseStorageSchema(schemaStr)
 	c.Assert(err, IsNil)
 
-	schemas, err := schema.SchemaAt([]string{"foo", "0"})
+	schemas, err := schema.SchemaAt([]string{"foo", "[0]"})
```
Comment (ZeyadYasser) [inline confdb/schema_test.go]: Is there a test for index placeholders e.g. `[{n}]`?

File: confdb/confdb.go [hunk 8]
```diff
@@ -505,14 +598,36 @@ func getPlaceholders(viewStr string) map[string]int {
...
 	}
 
-	return placeholders
+	return placeholders, nil
+}
+
+func getKeyPlaceholders(path string) (map[string]int, error) {
+	return getPlaceholders(path, func(key string) (string, bool) {
+		if isPlaceholder(key) {
+			return key[1 : len(key)-1], true
```
Comment (ZeyadYasser) [inline confdb/confdb.go]: a comment explaining what the filter does for key placeholder would be nice

File: confdb/confdb.go [hunk 9]
```diff
@@ -505,14 +598,36 @@ func getPlaceholders(viewStr string) map[string]int {
...
+		}
+		return "", false
+	})
+}
+
+func getIndexPlaceholders(path string) (map[string]int, error) {
+	return getPlaceholders(path, func(key string) (string, bool) {
+		if isIndexPlaceholder(key) {
+			return key[2 : len(key)-2], true
```
Comment (ZeyadYasser) [inline confdb/confdb.go]: same as above but for index placeholders

File: confdb/confdb.go [hunk 10]
```diff
@@ -873,6 +995,7 @@ func replaceIn(path, key, value string) string {
 		}
 	}
 
+	// TODO: what to do about this. will have to do this manually
 	return strings.Join(parts, ".")
```
Comment (ZeyadYasser) [inline confdb/confdb.go:1034]: would this break indices?

File: confdb/confdb.go [hunk 11]
```diff
@@ -1327,6 +1469,10 @@ func isPlaceholder(part string) bool {
 	return part[0] == '{' && part[len(part)-1] == '}'
 }
 
+func isIndexPlaceholder(part string) bool {
+	return len(part) > 4 && part[0] == '[' && part[len(part)-1] == ']' && isPlaceholder(part[1:len(part)-1])
```
Comment (ZeyadYasser) [inline confdb/confdb.go]: why not use the existing `validIndexPlaceholder` regex?

File: confdb/confdb.go [hunk 12]
```diff
@@ -1342,12 +1488,12 @@ type viewRule struct {
 // match returns true if the subkeys match the pattern exactly or as a prefix.
 // If placeholders are "filled in" when matching, those are returned in a map.
 // If the subkeys match as a prefix, the remaining suffix is returned.
-func (p *viewRule) match(reqSubkeys []string) (placeholders map[string]string, restSuffix []string, match bool) {
+func (p *viewRule) match(reqSubkeys []string) (placeholders *matchedPlaceholders, restSuffix []string, match bool) {
```
Comment (ZeyadYasser) [inline confdb/confdb.go]: doc comment needs to be updated to reflect using `matchedPlaceholders`

File: confdb/confdb.go [hunk 13]
```diff
@@ -1364,20 +1510,16 @@ func (p *viewRule) match(reqSubkeys []string) (placeholders map[string]string, r
 
 // storagePath takes a map of placeholders to their values in the view name and
 // returns the path with its placeholder values filled in with the map's values.
-func (p *viewRule) storagePath(placeholders map[string]string) (string, error) {
+func (p *viewRule) storagePath(matched *matchedPlaceholders) (string, error) {
```
Comment (ZeyadYasser) [inline confdb/confdb.go:1547]: doc comment needs to be updated to reflect using `matchedPlaceholders`

File: confdb/confdb.go (hunk 12 above)
Comment (ZeyadYasser) [reply confdb/confdb.go]:
```suggestion
func (p *viewRule) match(reqSubkeys []string) (matched *matchedPlaceholders, restSuffix []string, match bool) {
```
for consistency

File: confdb/confdb.go [hunk 14]
```diff
@@ -1408,18 +1554,28 @@ type placeholder string
 
 // match adds a mapping to the placeholders map from this placeholder key to the
 // supplied name subkey and returns true (a placeholder matches with any value).
-func (p placeholder) match(subkey string, placeholders map[string]string) bool {
-	placeholders[string(p)] = subkey
+func (p placeholder) match(subkey string, matched *matchedPlaceholders) bool {
```
Comment (ZeyadYasser) [inline confdb/confdb.go]: doc comment needs to be updated to reflect using `matchedPlaceholders`

File: confdb/confdb.go [hunk 15]
```diff
@@ -1408,18 +1554,28 @@ type placeholder string
...
+	}
+	matched.key[string(p)] = subkey
 	return true
 }
 
 // write writes the value from the placeholders map corresponding to this placeholder
 // key into the strings.Builder.
-func (p placeholder) write(sb *strings.Builder, placeholders map[string]string) error {
-	subkey, ok := placeholders[string(p)]
+func (p placeholder) write(sb *strings.Builder, matched *matchedPlaceholders, opts writeOpts) error {
```
Comment (ZeyadYasser) [inline confdb/confdb.go]: doc comment needs to be updated to reflect using `matchedPlaceholders`

File: confdb/confdb.go [hunk 16]
```diff
@@ -1431,16 +1587,74 @@ func (p placeholder) String() string {
...
+	key   map[string]string
+}
+
+// indexPlaceholder represents a subkey of a name/path (e.g., "[{n}]") that can
+// match an index value and map it from the input name to the path.
+type indexPlaceholder string
+
+// match checks if the subkey can be used to index a list. If so, it adds a
+// mapping from this placeholder key to the supplied name subkey and returns true.
+func (p indexPlaceholder) match(subkey string, matched *matchedPlaceholders) bool {
```
Comment (ZeyadYasser) [inline confdb/confdb.go:1634]: same as above regarding doc comment

File: confdb/confdb.go [hunk 17]
```diff
@@ -1431,16 +1587,74 @@ func (p placeholder) String() string {
...
+	if len(subkey) < 3 || !(subkey[0] == '[' && subkey[len(subkey)-1] == ']') {
+		return false
+	}
+
+	if _, err := strconv.Atoi(subkey[1 : len(subkey)-1]); err != nil {
+		// subkey can't be used as a index placeholder value
+		return false
+	}
```
Comment (ZeyadYasser) [inline confdb/confdb.go]: reuse regex?

File: confdb/confdb.go [hunk 18]
```diff
@@ -1431,16 +1587,74 @@ func (p placeholder) String() string {
...
+	if matched.index == nil {
+		matched.index = make(map[string]string)
+	}
+	matched.index[string(p)] = subkey[1 : len(subkey)-1]
+	return true
+}
+
+// write writes the value from the placeholders map corresponding to this placeholder
+// key into the strings.Builder.
+func (p indexPlaceholder) write(sb *strings.Builder, matched *matchedPlaceholders, opts writeOpts) error {
```
Comment (ZeyadYasser) [inline confdb/confdb.go]: same as above regarding doc comment

File: confdb/confdb.go (hunk 5 above)
Comment (ZeyadYasser) [reply confdb/confdb.go]: can't extraction be done directly with regex defined at the beginning, it would verify and extract?

File: confdb/confdb.go (hunk 4 above)
Comment (miguelpires) [reply confdb/confdb.go]: I didn't do that because {foo} is a subset of the index placeholder (we even call isPlaceholder from isIndexPlaceholder).
```
func isIndexPlaceholder(part string) bool {
	return len(part) > 4 && part[0] == '[' && part[len(part)-1] == ']' && isPlaceholder(part[1:len(part)-1])
}
```
So I thought it would be confusing to see IsIndexPlaceholder checking if a stripped part also matches IsKeyPlaceholder. So I left it as a generic name that applies to both types of placeholder

File: confdb/confdb.go (hunk 6 above)
Comment (miguelpires) [reply confdb/confdb.go:1602]: Yes but this is still be safe because accesses on nil maps return `"", false`

File: confdb/confdb.go (hunk 10 above)
Comment (miguelpires) [reply confdb/confdb.go:1034]: I just need to assemble the parts manually (since we now have "[]" accessors and not just ".") but it's not needed for this PR so I opted not to include. I'll include it in the next one (it adds the traversal logic)

File: confdb/schema_test.go (hunk 7 above)
Comment (miguelpires) [reply confdb/schema_test.go]: Good point, there wasn't for this specific function. I added one now, thanks

File: confdb/confdb.go (hunk 11 above)
Comment (miguelpires) [reply confdb/confdb.go]: We could, these are just tailored to a different usage. The validation uses the regex patterns because it needs to be strict and it's fine if its slow since it only happens once at creation time. These helpers are used to determine which kind of subkey we're looking at so they don't need to be as strict but they should (ideally) be fast

File: confdb/confdb.go (hunk 12 above)
Comment (miguelpires) [reply confdb/confdb.go]: done, thanks

File: confdb/confdb.go (hunk 5 above)
Comment (miguelpires) [reply confdb/confdb.go]: I can add a helper that does both checking and slicing, to couple the validation with the reslicing. We mostly do this in validation plus a couple of other places where we need to keep the [] around in user supplied paths (so we know which type of subkey we're looking at)

I could do the validation and extraction in the same place but then I'd have to pass those lists around out of the validation and into this code which IMO makes things more confusing. Currently, the path validation happens alongside other validation checks and can only return an error.

File: confdb/confdb.go (hunk 6 above)
Comment (andrewphelpsj) [reply confdb/confdb.go:1602]: Ah yeah, it is writes to a nil map that panic.

File: confdb/confdb.go (hunk 17 above)
Comment (miguelpires) [reply confdb/confdb.go]: used the helpers as mentioned in another comment

File: confdb/confdb.go [hunk 19]
```diff
@@ -470,42 +523,105 @@ type validationOptions struct {
...
 	}
 
 	return nil
 }
 
-// getPlaceholders returns the set of placeholders in the string or nil, if
-// there is none.
-func getPlaceholders(viewStr string) map[string]bool {
-	var placeholders map[string]bool
+func splitViewPath(path string) ([]string, error) {
```
Comment (pedronis) [inline confdb/confdb.go:600]: wouldn't it make sense to return something a bit more complex than string here, like an interface that can deal with index vs key more nicely?

**Review by pedronis — COMMENTED** [review]
high-level question

File: confdb/confdb.go (hunk 19 above)
Comment (miguelpires) [reply confdb/confdb.go:600]: Maybe but my concern is that this gets used all over the code, for different things. Just returning each a list of access sub-key works with all the use cases whereas an equivalent interface would have to have lots of methods for different use cases. But maybe you have something concrete in mind that doesn't have that issue?

Comment (pedronis) [reply confdb/confdb.go:600]: wouldn't the method always be the same, just be about is it a literal or placeholder? get the placeholder name? and get the key value or the index value?

Comment (miguelpires) [reply confdb/confdb.go:600]: ah so just moving the isIndex(), isPlaceholder(), etc methods into a type. I can do that, sure. The diff will be pretty large though since lots of code depends on these paths

Comment (pedronis) [reply confdb/confdb.go:600]: yes, but it also mean they can return a constant if we use different types and we do the recognizing when we split. Also you can add extra methods as well, if they have more than a couple usage places

File: confdb/confdb.go (hunk 11 above)
Comment (ZeyadYasser) [reply confdb/confdb.go]: makes sense, thanks for the explaination

**Review by ZeyadYasser — DISMISSED** [review]
Thank you!

File: confdb/confdb.go (hunk 19 above)
Comment (miguelpires) [reply confdb/confdb.go:600]: I pushed 3ee55bbf20b3121bdade232da4b8d229de52131c I'm not sure if that's what you had in mind. It can/will be extended to the traversal logic once this is merged

File: confdb/confdb.go [hunk 20]
```diff
@@ -388,7 +391,21 @@ func parseRule(parent *viewRule, ruleRaw any) ([]*viewRule, error) {
 		return nil, errors.New(`"request" must be a string`)
 	}
 
-	if err := validateRequestStoragePair(request, storage); err != nil {
+	// content sub-rules are shorthands for paths that include the parent's path
```
Comment (pedronis) [inline confdb/confdb.go:394]: I'm not sure I understand the reason for this new code from the comment

File: confdb/confdb.go [hunk 21]
```diff
@@ -465,47 +533,150 @@ type validationOptions struct {
...
+//     placeholders (e.g., foo.{bar}, a[{n}].bar), if allowed by the validationOptions
 //   - non-placeholder subkeys are made up of lowercase alphanumeric ASCII characters,
 //     optionally with dashes between alphanumeric characters (e.g., "a-b-c")
 //   - placeholder subkeys are composed of non-placeholder subkeys wrapped in curly brackets
-func validateViewDottedPath(path string, opts *validationOptions) (err error) {
+//   - bracketed subkeys that aren't placeholders can only contain integers
+//
+// If the validation succeeds, it returns an []accessor which contains typed
+// representations of each type of subkey (e.g., key placeholder, index, etc).
+func validateViewDottedPath(path string, opts *validationOptions) ([]accessor, error) {
```
Comment (pedronis) [inline confdb/confdb.go]: I'm slightly confused by how the work is distributed between this, splitViewPath and pathIntoAccessors ? as it seems the loop here could do what pathIntoAccessors is doing?

File: confdb/confdb.go [hunk 22]
```diff
@@ -691,7 +864,11 @@ func (v *View) Unset(databag Databag, request string) error {
 
 func (v *View) matchWriteRequest(request string) ([]requestMatch, error) {
 	var matches []requestMatch
-	subkeys := strings.Split(request, ".")
+	subkeys, err := splitViewPath(request)
```
Comment (pedronis) [inline confdb/confdb.go:851]: as mentioned elsewhere some of this should use something that produce accessors? can you leave at least TODO where you think it's the case?

File: confdb/confdb.go [hunk 23]
```diff
@@ -1387,62 +1574,148 @@ func (p viewRule) isWriteable() bool {
...
 // with any value and map it from the input name to the path.
-type placeholder string
+type keyPlaceholder string
 
-// match adds a mapping to the placeholders map from this placeholder key to the
+// match adds an entry to matchedPlaceholders mapping this placeholder key to the
 // supplied name subkey and returns true (a placeholder matches with any value).
-func (p placeholder) match(subkey string, placeholders map[string]string) bool {
-	placeholders[string(p)] = subkey
+func (p keyPlaceholder) match(subkey string, matched *matchedPlaceholders) bool {
```
Comment (pedronis) [inline confdb/confdb.go:1591]: will the match methods continue to take a string or use an accessor at some point?

**Review by pedronis — COMMENTED** [review]
some questions

File: confdb/confdb.go (hunk 22 above)
Comment (miguelpires) [reply confdb/confdb.go:851]: Yes, I didn't extend the accessors to cover the matching to prevent this from getting any bigger. I'll put TODOs here as wel

File: confdb/confdb.go (hunk 23 above)
Comment (miguelpires) [reply confdb/confdb.go:1591]: I think they'll use an accessor but we need the Get and Set changes in place first (since we can't change the interface for just one side)

File: confdb/confdb.go (hunk 20 above)
Comment (miguelpires) [reply confdb/confdb.go:394]: This is just taking the parent's path and prefixing it to deal with "content" rules. We previously did it later but since newViewRule now takes the accessors and converts them into matchers/writers, we have to build the nested path in the nested call and not just before recursing, like we used to.

File: confdb/confdb.go (hunk 21 above)
Comment (miguelpires) [reply confdb/confdb.go]: Initially I thought it might be necessary to have pathIntoAccessors as a standalone to use in other places. I've merged the accessor creation into this now

File: confdb/confdb.go [hunk 24]
```diff
@@ -1275,49 +1448,53 @@ func (v *View) matchGetRequest(request string) (matches []requestMatch, err erro
...
+
+	return sb.String()
+}
+
 func isPlaceholder(part string) bool {
-	return part[0] == '{' && part[len(part)-1] == '}'
+	return len(part) > 2 && part[0] == '{' && part[len(part)-1] == '}'
+}
+
+func stripIndex(part string) (string, bool) {
```
Comment (pedronis) [inline confdb/confdb.go:1501]: nitpick: shouldn't this be later after/closer to the usages?

File: confdb/confdb.go [hunk 25]
```diff
@@ -724,13 +885,16 @@ func checkSchemaMismatch(schema DatabagSchema, rules []*viewRule) error {
 out:
 	for _, rule := range rules {
 		path := rule.originalStorage
-		pathParts := strings.Split(path, ".")
+		pathParts, err := splitViewPath(path)
```
Comment (pedronis) [inline confdb/confdb.go:888]: this will stay as is?

File: confdb/confdb.go [hunk 26]
```diff
@@ -878,7 +1043,10 @@ func checkForUnusedBranches(value any, paths map[string]struct{}) error {
 		var err error
 		var pathParts []string
 		if path != "" {
-			pathParts = strings.Split(path, ".")
+			pathParts, err = splitViewPath(path)
```
Comment (pedronis) [inline confdb/confdb.go:1046]: same question

File: confdb/confdb.go (hunk 25 above)
Comment (miguelpires) [reply confdb/confdb.go:888]: Yes, this is a storage path. Like we talked about there isn't much benefit to converting the storage side of things and it propagates complexity to the databag and schema code

File: confdb/confdb.go (hunk 24 above)
Comment (miguelpires) [reply confdb/confdb.go:1501]: The next PR will extend this refactor to the matching and Get/Set/etc and this can be removed then

File: confdb/confdb.go [hunk 27]
```diff
@@ -432,30 +444,86 @@ func parseRule(parent *viewRule, ruleRaw any) ([]*viewRule, error) {
...
+// path (and vice-versa) and that we don't use them to mean more than one thing
+func checkForMatchingPlaceholders(request, storage string, reqPlaceholders, storagePlaceholders map[string]int) error {
 	if len(reqPlaceholders) != len(storagePlaceholders) {
 		return fmt.Errorf("request %q and storage %q have mismatched placeholders", request, storage)
 	}
 
-	for placeholder := range reqPlaceholders {
-		if !storagePlaceholders[placeholder] {
+	for name, count := range reqPlaceholders {
+		if count != 1 {
```
Comment (ZeyadYasser) [inline confdb/confdb.go:517]: is the same count check needed for storagePlaceholders?

File: confdb/confdb.go [hunk 28]
```diff
@@ -432,30 +444,86 @@ func parseRule(parent *viewRule, ruleRaw any) ([]*viewRule, error) {
 }
 
 // validateRequestStoragePair checks that:
-//   - request and storage are composed of valid subkeys (see: validateViewString)
+//   - request and storage are composed of valid subkeys (see: validateViewDottedPath)
```
Comment (ZeyadYasser) [inline confdb/confdb.go:447]:
```suggestion
//   - request and storage are composed of valid subkeys (see: parsePathIntoAccessors)
```

File: confdb/confdb.go [hunk 29]
```diff
@@ -1275,49 +1448,53 @@ func (v *View) matchGetRequest(request string) (matches []requestMatch, err erro
 
 func (v *View) ID() string { return v.schema.Account + "/" + v.schema.Name + "/" + v.Name }
 
-func newViewRule(request, storage, accesstype string) (*viewRule, error) {
+func newViewRule(request, storage []accessor, accesstype string) (*viewRule, error) {
```
Comment (ZeyadYasser) [inline confdb/confdb.go:1451]: Type assertions below feel error prone, why not pass the expected types directly doing the "safe" type assertions on the callers side?

```suggestion
func newViewRule(request []requestMatcher, storage []storageWriter, accesstype string) (*viewRule, error) {
```

Comment (ZeyadYasser) [reply confdb/confdb.go:1451]: Maybe I am missing something and it is safe, in that case a comment explaining why would be great

File: confdb/confdb.go (hunk 24 above)
Comment (ZeyadYasser) [inline confdb/confdb.go:1501]: doc comment explaining the return values would be nice

File: confdb/confdb.go [hunk 30]
```diff
@@ -1387,62 +1559,148 @@ func (p viewRule) isWriteable() bool {
...
+// match checks if the subkey can be used to index a list. If so, it adds an
+// entry to matchedPlaceholders mapping this placeholder key to the supplied
+// name subkey and returns true.
+func (p indexPlaceholder) match(subkey string, matched *matchedPlaceholders) bool {
+	subkey, ok := stripIndex(subkey)
+	if !ok {
+		return false
+	}
+
+	if _, err := strconv.Atoi(subkey); err != nil {
```
Comment (ZeyadYasser) [inline confdb/confdb.go:1640]: shouldn't index placeholders be ascii e.g. `[{n}]`? I might be confusing something else.

Comment (ZeyadYasser) [reply confdb/confdb.go:1640]: also seems to be missing a unit test for errors

**Review by ZeyadYasser — COMMENTED** [review]
some comments and questions

File: confdb/confdb.go (hunk 30 above)
Comment (miguelpires) [reply confdb/confdb.go:1640]: This is the matching of a path part to that index placeholder, so we expect the provided piece that matches to that placeholder to be an int. This is going to be dropped in the next PR. Basically I'll use the new types that implement `accessor` to simplify the matching and drop the string parsing we do here so that will address the coverage issue (or if it doesn't I'd rather address it there if that's ok)

File: confdb/confdb.go (hunk 24 above)
Comment (miguelpires) [reply confdb/confdb.go:1501]: I'm removing this in the next PR (the matching won't need it anymore because it'll use the new types that implement accessor)

File: confdb/confdb.go (hunk 29 above)
Comment (miguelpires) [reply confdb/confdb.go:1451]: I've improved the checks but I can't move this out of the method, I need the accessor methods to rebuild the original path (the rule requires it)
